service/forward_service.cc

```diff
@@ -365,6 +365,13 @@ future<query::forward_result> forward_service::dispatch_to_shards(
     });
 }
 
+static lowres_clock::time_point compute_timeout(const query::forward_request& req) {
+    lowres_system_clock::duration time_left = req.timeout - lowres_system_clock::now();
+    lowres_clock::time_point timeout_point = lowres_clock::now() + time_left;
+
+    return timeout_point;
```

Comment:
This is still vulnerable in that system_clock is not guaranteed to have the same baseline everywhere (due to clock skew). But this is a rare problem and requires misconfiguration.